### PR TITLE
[WIP] Native queue length metric for SQL transport

### PR DESCRIPTION
**Pull request by tmasternak**
The goal for this issue PR is to provide native queue length metric infrastructure and an implementation  for SQL transport. The general approach  is:
 * We discover endpoint input queues by means of `EndpointMetadataReport`s sent from endpoints,
 * If endpoint level monitoring is on the `TaggedLongValueOcurrence` reports will be sent as well,
 * Both of those reports are passed to `QueueLengthProvider` implementation. It will be `DefaultQueueLengthProvider` for those that support endpoint level monitoring,
 * Queue length provider is responsible for updating `QueueLengthStore` which holds queue length value per logical endpoint broken down by input queue name,
 * The overall metric value is calculated per logical endpoint and is the sum of average values per input queue,
* External API stays the same,

This PR requires https://github.com/Particular/NServiceBus.Metrics.ServiceControl/pull/34

### CODE REVIEW

**Comment by Scooletz** [comment]
@tmasternak Could you provide a description that shows the direction where is this PR going? It would be easier for a reviewer to "just review".

**Comment by tmasternak** [comment]
@Scooletz description upated.

**Comment by Scooletz** [comment]
Thank you for updating the description @tmasternak 
I skimmed through the PR again. Will review as soon as it's un-wipped.

**Comment by tmasternak** [comment]
This can be un-wiped when there is native queue length version of the plugin targeting core v6. We need this to smoke test the integration.

File: src/ServiceControl.Monitoring/EndpointFactory.cs [hunk 1]
```diff
@@ -76,12 +82,32 @@ public static void MakeMetricsReceiver(EndpointConfiguration config, Settings se
...
 
-        static IContainer CreateContainer(Settings settings)
+        static Func<QueueLengthStore, IProvideQueueLength> QueueLengthProviderBuilder(string explicitConnectionStringValue, Type transportType)
+        {
+            return qls =>
+            {
+                var connectionString = explicitConnectionStringValue ?? ConfigurationManager.ConnectionStrings["NServiceBus/Transport"]?.ConnectionString;
+
+                var queueLengthProviderType = transportType.Assembly.GetTypes()
+                    .SingleOrDefault(p => typeof(IProvideQueueLength).IsAssignableFrom(p));
```
[the remark this answers is not in the document]
Comment (mikeminutillo) [reply src/ServiceControl.Monitoring/EndpointFactory.cs]: I think we'll require every transport to have an SC-specific implementation. Even if all it does it inherit from the "real" version. That will allow every transport to provide a Queue Length Provider

File: src/ServiceControl.Monitoring/QueueLength/LegacyQueueLengthReportHandler.cs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    using NServiceBus.Logging;
+    using NServiceBus.Metrics;
+
+    class LegacyQueueLengthReportHandler : IHandleMessages<MetricReport>
+    {
+        public Task Handle(MetricReport message, IMessageHandlerContext context)
+        {
+            var endpointInstanceId = EndpointInstanceId.From(context.MessageHeaders);
+
+            Logger.Warn($"Legacy queue length report received from {endpointInstanceId.InstanceName} instance of {endpointInstanceId.EndpointName}");
```
[the remark this answers is not in the document]
Comment (mikeminutillo) [reply src/ServiceControl.Monitoring/QueueLength/LegacyQueueLengthReportHandler.cs]: I suspect we'll want to track this and show it in the UI somewhere.

**Comment by mikeminutillo** [comment]
@Scooletz would it be valuable to split the SQL-specific implementation from the rest of the changes? If we can merge the architectural changes to develop then we can start on the other transports.

**Comment by mikeminutillo** [comment]
Split in to #121 (infra) and #122 (sql-specific)

**Pull request closed without merging**